### PR TITLE
[improve][broker] PIP-383: Support granting/revoking permissions for multiple topics

### PR DESCRIPTION
**Pull request by Technoboy-**
### Motivation

PIP-383:  https://github.com/apache/pulsar/pull/23355

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java [hunk 1]
```diff
@@ -251,6 +254,68 @@ public CompletableFuture<Void> grantPermissionAsync(TopicName topicName, Set<Aut
...
+        return getPoliciesReadOnlyAsync().thenCompose(readonly -> {
+            if (readonly) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Policies are read-only. Broker cannot do read-write operations");
+                }
+                throw new IllegalStateException("policies are in readonly mode");
+            }
+            TopicName topicName = TopicName.get(options.get(0).getTopic());
+            return pulsarResources.getNamespaceResources()
+                    .setPoliciesAsync(topicName.getNamespaceObject(), policies -> {
```
Comment (lhotari) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java:267]: This code seems to assume that all topics are in the same namespace. That seems like a problematic assumption. Is there a goal to support `GrantTopicPermissionOptions` that apply to topics that could be in different namespaces?

Although I can see that there's a check in the different layer. The validation should happen here again unless support for multiple namespaces is added. The reason for this is that the caller shouldn't need to know implementation details. If it needs to know the implementation details, it's a leaky or broken abstraction.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java [hunk 2]
```diff
@@ -251,6 +254,68 @@ public CompletableFuture<Void> grantPermissionAsync(TopicName topicName, Set<Aut
...
```
Comment (lhotari) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java:296]: This code seems to assume that all topics are in the same namespace. Similar comment as in the previous case.

**Review by lhotari — CHANGES_REQUESTED** [review]

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java (hunk 1 above)
Comment (Technoboy-) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java:267]: yes, Lari, should be in the same namespace.
I have checked it in the Rest Part

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java [hunk 3]
```diff
@@ -613,6 +616,83 @@ protected CompletableFuture<Void> internalGrantPermissionOnNamespaceAsync(String
...
+                        throw new RestException(Status.CONFLICT, "Concurrent modification");
+                    } else {
+                        log.error("[{}] Failed to revoke permissions for namespace {}",
+                                clientAppId(), namespaceName, ex);
+                        throw new RestException(realCause);
+                    }
+                });
+    }
+
+    private CompletableFuture<Boolean> checkNamespace(Stream<String> namespaces) {
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:686]: I have checked it here

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java (hunk 1 above)
Comment (Technoboy-) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java:267]: https://github.com/apache/pulsar/pull/23372/files#r1794452230
@lhotari

Comment (Technoboy-) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java]: Thanks @lhotari , I have addressed your comment. add the validation here

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java (hunk 2 above)
Comment (Technoboy-) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authorization/PulsarAuthorizationProvider.java]: addressed

**Review by lhotari — APPROVED** [review]
LGTM, good work @Technoboy-

**Pull request merged**